lib/async_hooks.js

```diff
@@ -0,0 +1,486 @@
+  }
+}
+
+
+// Sensitive Embedder API //
+
+// Increment the internal id counter and return the value. Important that the
+// counter increment first. Since it's done the same way in
+// Environment::new_async_uid()
+function newUid() {
+  return ++async_uid_fields[kAsyncUidCntr];
+}
+
+
+// Return the triggerId meant for the constructor calling it. It's up to the
+// user to safeguard this call and make sure it's zero'd out when the
+// constructor is complete.
+function initTriggerId() {
+  var tId = async_uid_fields[kInitTriggerId];
+  // Reset value after it's been called so the next constructor doesn't
+  // inherit it by accident.
+  async_uid_fields[kInitTriggerId] = 0;
+  if (tId <= 0)
+    tId = async_uid_fields[kCurrentAsyncId];
+  return tId;
+}
+
+
+function setInitTriggerId(triggerId) {
+  // CHECK(Number.isSafeInteger(triggerId))
+  // CHECK(triggerId > 0)
+  async_uid_fields[kInitTriggerId] = triggerId;
+}
+
+
+function emitInitS(asyncId, type, triggerId, resource) {
+  // Short circuit all checks for the common case. Which is that no hooks have
+  // been set. Do this to remove performance impact for embedders (and core).
+  // Even though it bypasses all the argument checks. The performance savings
+  // here is critical.
+  if (async_hook_fields[kInit] === 0)
+    return;
+
+  // This can run after the early return check b/c running this function
+  // manually means that the embedder must have used initTriggerId().
+  if (!Number.isSafeInteger(triggerId)) {
+    if (triggerId !== undefined)
+      resource = triggerId;
+    triggerId = initTriggerId();
+  }
+
+  // I'd prefer allowing these checks to not exist, or only throw in a debug
+  // build, in order to improve performance.
+  if (!Number.isSafeInteger(asyncId) || asyncId < 0)
+    throw new RangeError('asyncId must be an unsigned integer');
+  if (typeof type !== 'string' || type.length <= 0)
+    throw new TypeError('type must be a string with length > 0');
+  if (!Number.isSafeInteger(triggerId) || triggerId < 0)
+    throw new RangeError('triggerId must be an unsigned integer');
+
+  for (var i = 0; i < active_hooks_array.length; i++) {
+    if (typeof active_hooks_array[i][init_symbol] === 'function') {
+      runInitCallback(
+        active_hooks_array[i][init_symbol], asyncId, type, triggerId, resource);
+    }
+  }
+
+  // Isn't null if hooks were added/removed while the hooks were running.
+  if (tmp_active_hooks_array !== null) {
+    restoreTmpHooks();
+  }
+}
+
+
+function emitBeforeN(asyncId, triggerId) {
+  for (var i = 0; i < active_hooks_array.length; i++) {
+    if (typeof active_hooks_array[i][before_symbol] === 'function') {
+      runCallback(active_hooks_array[i][before_symbol], asyncId);
+    }
+  }
+
+  if (tmp_active_hooks_array !== null) {
+    restoreTmpHooks();
+  }
+}
+
+
+// Usage: emitBeforeS(asyncId[, triggerId]). If triggerId is omitted then
+// asyncId will be used instead.
+function emitBeforeS(asyncId, triggerId) {
```

Comment:
> A `triggerId === 0` has a specific meaning and would take a different logic path than `triggerId === undefined`.

I commented with that because I didn’t find that different path… am I mistaken? Or is it because of that triggerId-is-unused-in-emitBeforeN-but-should-be-used thing above?